### ISSUE TITLE
feat: Mixup and CutMix regularization for training

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#210]
## Summary

Mixup and CutMix are data-level regularization techniques that significantly improve ImageNet accuracy. They work by blending training samples and their labels, forcing the model to learn smoother decision boundaries.

**Estimated impact**: ~1–2% Top-1 accuracy improvement on ImageNet.

### Mixup (Zhang et al., 2018)
Blends two random training samples:
```
λ ~ Beta(α, α)      // e.g., α = 0.2
x_mixed = λ * x_i + (1 - λ) * x_j
y_mixed = λ * y_i + (1 - λ) * y_j     // soft labels!
```

### CutMix (Yun et al., 2019)
Cuts a rectangular patch from one image and pastes it onto another:
```
λ ~ Beta(α, α)
(x0, y0, w, h) = random_box(image_size, λ)
x_mixed = x_i  with  x_mixed[:, y0:y0+h, x0:x0+w] = x_j[:, y0:y0+h, x0:x0+w]
y_mixed = λ_actual * y_i + (1 - λ_actual) * y_j
where λ_actual = 1 - (w * h) / (W * H)   // actual pixel ratio
```

### Mixup

```java
public final class Mixup {
    private final float alpha;  // Beta distribution parameter (default 0.2)
    
    public Mixup(float alpha) { this.alpha = alpha; }
    
    /**
     * Applies Mixup to a batch of data and labels.
     * Shuffles the batch internally and blends with random λ.
     *
     * @param data   [B, C, H, W] batch tensor
     * @param labels [B, numClasses] one-hot labels
     * @return MixResult with blended data and soft labels
     */
    public MixResult apply(Tensor data, Tensor labels) { ... }
    
    public record MixResult(Tensor data, Tensor labels) {}
}
```

### CutMix

```java
public final class CutMix {
    private final float alpha;
    
    public CutMix(float alpha) { this.alpha = alpha; }
    
    public MixResult apply(Tensor data, Tensor labels) { ... }
}
```

### Combined Usage (common practice: randomly choose Mixup or CutMix per batch)

```java
public final class MixupCutMix {
    private final Mixup mixup;
    private final CutMix cutmix;
    private final float mixupProb;  // probability of choosing Mixup vs CutMix
    
    public MixupCutMix(float alpha, float mixupProb) {
        this.mixup = new Mixup(alpha);
        this.cutmix = new CutMix(alpha);
        this.mixupProb = mixupProb;
    }
    
    public MixResult apply(Tensor data, Tensor labels) {
        if (ThreadLocalRandom.current().nextFloat() < mixupProb) {
            return mixup.apply(data, labels);
        } else {
            return cutmix.apply(data, labels);
        }
    }
}
```

### Integration with Trainer

```java
var trainer = new Trainer(model, loss, optimizer)
    .withMixup(new MixupCutMix(0.2f, 0.5f));  // 50% Mixup, 50% CutMix

// In training loop:
for (var batch : trainLoader) {
    var mixed = mixup.apply(batch.data(), batch.labels());
    var predictions = model.forward(mixed.data(), true);
    var lossResult = loss.compute(predictions, mixed.labels());  // soft label CE
    // ...
}
```

### Beta Distribution Sampling
Java doesn't have a built-in Beta distribution. Two approaches:
1. Use Gamma distribution: `Beta(α, β) = Gamma(α, 1) / (Gamma(α, 1) + Gamma(β, 1))`
2. Simple rejection sampling for small α values
3. Or add Apache Commons Math dependency (probably overkill)

### Soft Labels with CrossEntropyLoss
Current `CrossEntropyLoss` already supports soft labels since it computes:
```
loss = -Σ target[i] * log(prob[i])
```
This works correctly when targets sum to 1.0 but aren't one-hot.

### Batch-Level Operation
Mixup/CutMix operates on **entire batches**, not individual samples. This is different from augmentations like `RandomHorizontalFlip` that operate per-sample.

## Acceptance Criteria

- [ ] `Mixup` correctly blends batch samples with Beta-distributed λ
- [ ] `CutMix` correctly cuts and pastes rectangular regions
- [ ] Labels are soft-blended proportional to the mix ratio
- [ ] `MixupCutMix` randomly selects between the two per batch
- [ ] Beta distribution sampling implemented without external dependencies
- [ ] Works with `CrossEntropyLoss` (soft label support already exists)
- [ ] Thread-safe
- [ ] Unit test: Mixup with λ=0.5 produces exact average of two samples
- [ ] Unit test: CutMix actual λ matches pixel ratio
- [ ] Unit test: soft labels sum to 1.0

## References

- [Mixup (Zhang et al., 2018)](https://arxiv.org/abs/1710.09412)
- [CutMix (Yun et al., 2019)](https://arxiv.org/abs/1905.04899)
- [DeiT Training Recipe](https://arxiv.org/abs/2012.12877) — Uses both Mixup + CutMix
